### PR TITLE
Infographics

### PR DESCRIPTION
**Pull request by benghaziboy**
@mparent61 @MichaelButkovic 

Hey fellas, this is nowhere near ready...Just opening it up for y'all to get your eyes on it. Still working on how I want to separate out the data logic into a mixin, but kind of figuring out the pain points as I go along. 

--toodles.

### CODE REVIEW

**Comment by benghaziboy** [comment]
Thoughts are obviously requested

File: bulbs/contributions/models.py [hunk 1]
```diff
@@ -28,6 +29,13 @@ def calculate_hourly_pay(rate, minutes_worked):
     return ((float(rate) / 60) * minutes_worked)
 
 
+@shared_task(default_retry_delay=5)
+def updated_role_payment_type(role_pk):
```
Comment (MichaelButkovic) [inline bulbs/contributions/models.py]: did you intend for this to get included?

**Comment by mparent61** [comment]
Overall looking good. 
I see lots of seemingly unrelated changes in this PR too -- did you forget a merge?

**Comment by benghaziboy** [comment]
@MichaelButkovic @mparent61

File: setup.py [hunk 2]
```diff
@@ -19,8 +19,10 @@
     "beautifulsoup4>=4.4.1",
     "celery==3.1.10",
     "contextdecorator==0.10.0",
+    "django-enumfield==1.2.1",
     "django-betty-cropper>=0.2.6",
     "django-filter==0.9.2",
+    "django-jsonfield==0.9.19",
```
Comment (mparent61) [inline setup.py]: do we need both of these json field packages?

Comment (benghaziboy) [inline setup.py]: ```https://github.com/theonion/django-bulbs/blob/master/bulbs/ads/models.py#L7```

Uses the previous one that doesn't support `jsonb` I suggest we migrate that one to the newer package in another PR, since historically 3rd party field packages have been not terrific at tricky migrations.

File: bulbs/ads/apps.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import django.apps
```
Comment (MichaelButkovic) [inline bulbs/ads/apps.py]: i thought we nixed these?

Comment (mparent61) [inline bulbs/ads/apps.py]: We did nix them, branch needs a merge?

File: setup.py (hunk 2 above)
Comment (mparent61) [inline setup.py]: Fair enough, should add a TODO comment

File: bulbs/ads/apps.py (hunk 3 above)
Comment (benghaziboy) [inline bulbs/ads/apps.py]: previous merge gone awry

**Comment by benghaziboy** [comment]
fellas

**Comment by MichaelButkovic** [comment]
doesn't look like your merge worked

**Comment by benghaziboy** [comment]
check the last commit, the main comment page isn't indexed correctly.

**Comment by MichaelButkovic** [comment]
there's still app configs for the other apps, is what i'm saying

**Comment by benghaziboy** [comment]
ah feck

**Comment by benghaziboy** [comment]
ok players

File: CHANGELOG.md [hunk 4]
```diff
@@ -9,6 +9,7 @@
 ## Version 2.8.1
 
 - Reverted `AppConfig` changes released in 2.8.0. Decided to go another route with duplicate app names.
+>>>>>>> 43683584ffc91191e1dc4fd9d3cc06fb149b253a
```
Comment (MichaelButkovic) [inline CHANGELOG.md]: can remove

Comment (mparent61) [inline CHANGELOG.md]: gross

File: bulbs/api/views.py [hunk 5]
```diff
@@ -53,10 +54,15 @@
 
 
 class ContentViewMetaData(BaseMetadata):
-    def determine_metadata(self, request, view):
-        return {
-            'status': 'ok',
-        }
+
+    pass
+    # def determine_metadata(self, request, view):
```
Comment (MichaelButkovic) [inline bulbs/api/views.py]: not sure if you wanna just get rid of this, but hey

Comment (mparent61) [inline bulbs/api/views.py]: delete me or add a TODO

File: bulbs/contributions/tasks.py [hunk 6]
```diff
@@ -16,7 +16,7 @@ def check_and_update_freelanceprofiles(content_id):
 
 @shared_task(default_retry_delay=5)
 def update_role_rates(contributor_role_pk):
-    for contribution in Contribution.objects.filter(contributor__pk=contributor_role_pk):
+    for contribution in Contribution.objects.filter(role__pk=contributor_role_pk):
```
Comment (MichaelButkovic) [inline bulbs/contributions/tasks.py]: seems unrelated?

Comment (benghaziboy) [inline bulbs/contributions/tasks.py]: a bug that was caught due to a race condition during these changes

File: bulbs/infographics/metadata.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+from djbetty.serializers import ImageFieldSerializer
+
+from .data_serializers import CopySerializer, ItemSerializer, XYItemSerializer
+from .fields import ColorField, RichTextField
+from .serializers import BaseInfographicSerializer, InfographicDataField
+
+
+def get_and_check_attribute(obj, attr_name):
+    attribute = getattr(obj, attr_name)
```
Comment (mparent61) [inline bulbs/infographics/metadata.py]: Need to pass default param or this will raise.

File: bulbs/contributions/tasks.py (hunk 6 above)
Comment (MichaelButkovic) [inline bulbs/contributions/tasks.py]: gotcha

File: bulbs/infographics/metadata.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+    def get_serializer_info(self, serializer):
+        serializer_info = super(InfographicMetadataMixin, self).get_serializer_info(serializer)
+
+        if hasattr(serializer, "child"):
+            label = self.label_lookup[serializer.child]
+        else:
+            label = self.label_lookup[serializer]
+
+        # MIKE PARENT PAY ATTENTION: IDK IF THIS IS PRUDENT. LET'S DISCUSS.
```
Comment (mparent61) [inline bulbs/infographics/metadata.py]: paying attention...

File: bulbs/infographics/enum.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+from django_enumfield import enum
+
+
+class InfographicType(enum.Enum):
```
Comment (MichaelButkovic) [inline bulbs/infographics/enum.py]: I feel like this should just be in the `models.py` file

File: bulbs/infographics/models.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+    def clean(self, *args, **kwargs):
+        self.validate_data_field()
+        return super(BaseInfographic, self).clean(*args, **kwargs)
+
+    def get_data_serializer(self):
+        return get_data_serializer(self.infographic_type)
+
+    def validate_data_field(self):
+        Serializer = self.get_data_serializer()  # NOQA
```
Comment (mparent61) [inline bulbs/infographics/models.py]: why NOQA?

Comment (benghaziboy) [inline bulbs/infographics/models.py]: defining a localized variable with uppercase

File: example/settings.py [hunk 11]
```diff
@@ -7,8 +7,12 @@
 
 DATABASES = {
     'default': {
-        'ENGINE': 'django.db.backends.sqlite3',
-        'NAME': ':memory:'
+        'ENGINE': 'django.db.backends.postgresql_psycopg2',
```
Comment (mparent61) [inline example/settings.py]: ❤️

File: bulbs/infographics/enum.py (hunk 9 above)
Comment (benghaziboy) [inline bulbs/infographics/enum.py]: There are circular dependency issues regarding utility/validation between rest_framework + models.py that prohibit that. +++++ I think custom fields should be moved out of models anyway

File: bulbs/api/views.py [hunk 12]
```diff
@@ -75,7 +81,7 @@ class ContentViewSet(UncachedResponse, viewsets.ModelViewSet):
         "authors", "types"
     )
     permission_classes = [IsAdminUser, CanEditContent]
-    metadata_class = ContentViewMetaData
+    metadata_class = InfographicMetadataMixin
```
Comment (mparent61) [inline bulbs/api/views.py]: Why does a generic "ContentViewSet" have infographic-specific metadata? Or temp testing?

File: tests/infographics/test_api.py [hunk 13]
```diff
@@ -0,0 +1,427 @@
...
+                        ])
+                    )
+                ]),
+            }
+        })
+
+    def get_detail_endpoint(self, pk):
+        return reverse("content-detail", kwargs={"pk": pk})
+
+    @property
```
Comment (MichaelButkovic) [inline tests/infographics/test_api.py]: all of these properties are making these tests hard to read. not necessarily something that has to be fixed though.

File: bulbs/api/views.py (hunk 12 above)
Comment (benghaziboy) [inline bulbs/api/views.py]: Ah. this is something we should change, but it's a product of our polymorphic ContentView hierarchy.

Comment (benghaziboy) [inline bulbs/api/views.py]: could stand a refactyor

**Comment by MichaelButkovic** [comment]
👍

File: tests/infographics/test_api.py (hunk 13 above)
Comment (benghaziboy) [inline tests/infographics/test_api.py]: Didn't want to rewrite the same values over and over, the alternative is write them in the setUp, which yields the same behavior per test.

File: bulbs/api/views.py (hunk 12 above)
Comment (mparent61) [inline bulbs/api/views.py]: This feels like it's making it more complicated, because we're adding a content-specific metadata class to a base view, without much explanation. 
Why not move some of the `InfographicMetadataMixin.determine_metadata()` logic into a restored `InfographicMetadataMixin.determine_metadata()` method, at least then you can document the ugliness directly inside that method.

Comment (mparent61) [inline bulbs/api/views.py]: Also if i'm way off base let's just talk about it

Comment (benghaziboy) [inline bulbs/api/views.py]: well I think the solution (whilst we have this dynamic view stuff), is to have some generic way of registering metadata info as necessary.

**Comment by benghaziboy** [comment]
@mparent61 https://github.com/theonion/django-bulbs/pull/215/commits/91f477c7730a84e85f1f1d058abb514e644f1b99
is what I have decided to do for the time being.

**Comment by MichaelButkovic** [comment]
👍

File: bulbs/api/metadata.py [hunk 14]
```diff
@@ -0,0 +1,20 @@
+from rest_framework.metadata import SimpleMetadata
+
+from bulbs.infographics.metadata import InfographicMetadata
+from bulbs.infographics.serializers import InfographicSerializer
+
+
+class PolymorphicContentMetadata(SimpleMetadata):
```
Comment (mparent61) [inline bulbs/api/metadata.py]: :)

File: bulbs/infographics/metadata.py [hunk 15]
```diff
@@ -68,16 +66,17 @@ def get_serializer_info(self, serializer):
             ])
         return serializer_info
 
-    def get_custom_metadata(self, serializer, view):
+    def get_custom_metadata(self, serializer, view, list=False):
```
Comment (mparent61) [inline bulbs/infographics/metadata.py]: `list` isn't used anywhere...

File: bulbs/infographics/metadata.py [hunk 16]
```diff
@@ -68,16 +66,17 @@ def get_serializer_info(self, serializer):
...
         fields_metadata = dict()
         if hasattr(serializer, "__call__"):
             serializer_instance = serializer()
         else:
             serializer_instance = serializer
         for field_name, field in serializer_instance.get_fields().items():
             if isinstance(field, InfographicDataField):
-                serializer = view.get_object().get_data_serializer()
-                fields_metadata[field_name] = self.get_custom_metadata(serializer, view)
+                if view.get_view_name() != "Content List":
```
Comment (benghaziboy) [inline bulbs/infographics/metadata.py]: The only alternative I'm seeing atm is

```view.suffix == "List"``` unless we override the options, since we are trying to verify the origin of the request before the serializer is instantiated.

**Comment by mparent61** [comment]
All right, LGTM
